cellmlmanip/mathml2sympy/transpiler.py

```diff
@@ -1,355 +1,404 @@
-    return _wrapped_power
-
-
-def root_handler(node):
-    """
-    https://www.w3.org/TR/MathML2/chapter4.html#contm.root
-    operator taking qualifiers
-
-    Nasty:
-    The root element is used to construct roots. The kind of root to be taken is specified by a
-    degree element, which should be given as the second child of the apply element enclosing the
-    root element. Thus, square roots correspond to the case where degree contains the value 2, cube
-    roots correspond to 3, and so on. If no degree is present, a default value of 2 is used.
-    """
-    def _wrapped_root(first_argument, second_argument=None):
-        # if no <degree> given, it's sqrt
-        if second_argument is None:
-            return sympy.root(first_argument, 2)
-        return sympy.root(second_argument, first_argument)
-    return _wrapped_root
-
-
-def degree_handler(node):
-    """
-    https://www.w3.org/TR/MathML2/chapter4.html#contm.degree
-    Meaning of <degree> depends on context! We implement it for order of <bvar> in <diff> and
-    the kind of root in <root>
-    """
-    result = transpile(node)
-    if len(result) != 1:
-        raise ValueError('Expected single value in <degree> tag.'
-                         'Got: ' + node.toxml())
-    return result[0]
-
-
-# CALCULUS AND VECTOR CALCULUS #################################################################
-
-def diff_handler(node):
-    """
-    https://www.w3.org/TR/MathML2/chapter4.html#contm.diff
-    operator taking qualifiers
-    """
-    def _wrapped_diff(x_symbol, y_symbol, evaluate=False):
-        # dx / dy
-        y_function = sympy.Function(y_symbol.name)
-
-        # if bound variable element <bvar> contains <degree>, argument x_symbol is a list,
-        # otherwise, it is a symbol
-        if isinstance(x_symbol, list) and len(x_symbol) == 2:
-            bound_variable = x_symbol[0]
-            order = int(x_symbol[1])
-            return sympy.Derivative(y_function(bound_variable), bound_variable, order,
-                                    evaluate=evaluate)
-
-        return sympy.Derivative(y_function(x_symbol), x_symbol, evaluate=evaluate)
-    return _wrapped_diff
-
-
-def bvar_handler(node):
-    """
-    https://www.w3.org/TR/MathML2/chapter4.html#contm.bvar
-    NASTY: bvar element depends on the context it is being used
-    In a derivative, it indicates the variable with respect to which a function is being
-    differentiated.
-
-    The bound variable <bvar> can also specify degree. In this case, we'll have two elements
-    """
-    result = transpile(node)
-    if len(result) == 1:
-        # Bound variable without specifying degree
+                raise NotImplementedError('Unimplemented type attribute for <cn>: '
+                                          + node.attributes['type'].value)
+        else:
+            number = float(node.childNodes[0].data.strip())
+            number = sympy.Number(number)
+
+        if self.dummify:
```

Comment:
Do we use both options in the code? (With and without self.dummify) ? Otherwise maybe this option can go

Reply:
It's used in unit tests for mathml2sympy (outside the context of cellml models)